### PR TITLE
Fix cross-unit UDT dependency

### PR DESCRIPTION
**Pull request by swernli**
This fixes a subtle bug in usage of UDTs across compilation units. Because a UDT is captured as a type that includes a `Res` in it, it wil have the placeholder `None` for it's package id when initially compiled (package ids are only known after insertion into the store). Later, a type mismatch could occur because the caller has the package id in the UDT type but the signature stored in the scheme uses UDT types with `None`. This fix adds a notion of updating a scheme to include the package id at type checking time by using `with_package` such that type checking for both internal/`None` resolutions and dependencies from other compilation units with known package ids can all be properly used.

### CODE REVIEW

File: compiler/qsc_frontend/src/compile/tests.rs [hunk 1]
```diff
@@ -478,12 +482,12 @@ fn package_dependency() {
         None,
     );
     let unit2 = compile(&store, &[package1], sources2, TargetProfile::Full);
+    assert!(unit2.errors.is_empty(), "{:#?}", unit2.errors);
 
-    let foo_id = LocalItemId::from(1);
```
Comment (DmitryVasilevsky) [inline compiler/qsc_frontend/src/compile/tests.rs:482]: Is this change needed just to uncouple two blocks or LocalItemId::from(1) actually changes between two uses?

Comment (swernli) [reply compiler/qsc_frontend/src/compile/tests.rs:482]: Nope, the id is the same in both places. I was just removing it because I found it confusing more than helpful and it matches the pattern of the new test I added, but could revert if you think it's worth keeping. The ID does indeed stay the same between the two uses and could be a variable.

File: compiler/qsc_hir/src/ty.rs [hunk 2]
```diff
@@ -50,6 +50,25 @@ pub enum Ty {
...
     pub const UNIT: Self = Self::Tuple(Vec::new());
+
+    #[must_use]
+    pub fn with_package(&self, package: Option<PackageId>) -> Self {
+        match self {
+            Ty::Array(item) => Ty::Array(Box::new(item.with_package(package))),
+            Ty::Arrow(arrow) => Ty::Arrow(Box::new(arrow.with_package(package))),
+            Ty::Infer(infer) => Ty::Infer(*infer),
+            Ty::Param(param) => Ty::Param(*param),
+            Ty::Prim(prim) => Ty::Prim(*prim),
```
Comment (DmitryVasilevsky) [inline compiler/qsc_hir/src/ty.rs:61]: Don't know much Rust, hence the question. Can you use the same trick for the three cases as you used before ("_ => *self,")

Comment (swernli) [reply compiler/qsc_hir/src/ty.rs:61]: Yes, actually I could!

Comment (swernli) [reply compiler/qsc_hir/src/ty.rs:61]: Actually, no, because each case needs to construct a different type.

**Review by idavis — APPROVED** [review]

File: compiler/qsc_hir/src/ty.rs (hunk 2 above)
Comment (swernli) [reply compiler/qsc_hir/src/ty.rs:61]: Ok, never mind. Came back again, and you are right. Each case was created an instance of the same type, so those can all be captured by a single case that clones. Good eye!
